controllers/reconcile_persistence.go

```diff
@@ -0,0 +1,177 @@
+package controllers
+
+import (
+	"errors"
+	"context"
+	"fmt"
+	"github.com/go-logr/logr"
+	rabbitmqv1beta1 "github.com/rabbitmq/cluster-operator/api/v1beta1"
+	"github.com/rabbitmq/cluster-operator/internal/resource"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	k8sresource "k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+
+func (r *RabbitmqClusterReconciler) reconcilePVC(ctx context.Context, builder resource.ResourceBuilder, cluster *rabbitmqv1beta1.RabbitmqCluster, resource client.Object) error {
+	logger := ctrl.LoggerFrom(ctx)
+
+	switch sts := resource.(type) {
+	case *appsv1.StatefulSet:
+		current, err := r.statefulSet(ctx, cluster)
+		if client.IgnoreNotFound(err) != nil {
+			return err
+		} else if k8serrors.IsNotFound(err) {
+			logger.Info("statefulSet not created yet, skipping checks to expand PersistentVolumeClaims")
+			return nil
+		}
+
+		if err := builder.Update(sts); err != nil {
+			return err
+		}
+
+		resize, err := r.needsPVCResize(current, sts)
+		if err != nil {
+			return err
+		}
+
+		if resize {
+			if err := r.expandPVC(ctx, cluster, current, sts); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (r *RabbitmqClusterReconciler) expandPVC(ctx context.Context, rmq *rabbitmqv1beta1.RabbitmqCluster, current, desired *appsv1.StatefulSet) error {
+	logger := ctrl.LoggerFrom(ctx)
+
+	currentCapacity, err := persistenceStorageCapacity(current.Spec.VolumeClaimTemplates)
+	if err != nil {
+		return err
+	}
+
+	desiredCapacity, err := persistenceStorageCapacity(desired.Spec.VolumeClaimTemplates)
+	if err != nil {
+		return err
+	}
+
+	logger.Info(fmt.Sprintf("updating storage capacity from %s to %s", currentCapacity.String(), desiredCapacity.String()))
+
+	if err := r.deleteSts(ctx, rmq); err != nil {
+		return err
+	}
```

Comment:
Given that we have the `current` StatefulSet here (as argument) and inside `deleteSts()` we only use `rmq` to generate the name of the StatefulSet, to then run a `client.Get()` to get the same StatefulSet we have here (as `current`), perhaps we should pass just `current` StatefulSet instead of `rmq` since it has all the information we need and we can save one `client.Get()`.

Reply:
Sure! Good catch.